### DISCUSSION TITLE
sendParent throws exception if machine is not invoked

### DISCUSSION THREAD

**Discussion by kadeer** [statelyai/xstate#2047] [General]
Hi, is it possible to have XState not throw an exception when a machine that's not invoked but has sendParent actions fires? I'm also open to better ways to do what I'm attempting to do. For further context:

I have defined 2 machines that handles often paired components within my application. 

- Event machine: for managing the view states of an event in the system
- NoteEditor machine: for managing view states for notes in the system

A note can exist by itself in which case the NoteEditor machine is used to handle the state. A note can also be attached to an event, kind of like a comment, in which case I built an "Event+Note" composite machine that invokes the Event and NoteEditor machines to handle both Event and NoteEditor views. It's also useful because when notes are linked to an event, context menu items for its notes can be shown in the event.

In the Event+Note machine, I do some forwarding & sendParent action usage so that the API is cleaner for composite views. Because of the sendParent actions, I'm limited in using Note view + NoteEditorMachine by itself, as I get the following error when I create the machine as the root machine:

Error: Unable to send event to child '#_parent' from service 'createNote'

A few options I've considered:
- Using both machines individually and abandon the composite machine, but I have event-focused views, note-focused views, and views that manage both views and notes together, and it'd be nice not to repeat myself.
- Use an event or the context to store whether each child machine is invoked, and then invoke the sendParent action conditionally.

**Comment by davidkpiano** [comment]
✏️ Please create a [CodeSandbox (XState template)](https://codesandbox.io/s/xstate-example-template-qcwed) reproduction of this. Thanks!

**Reply by kadeer** [reply]
Hi @davidkpiano, thank you for your response. Here a repro: https://codesandbox.io/s/xstate-example-template-vanilla-forked-52430?file=/src/index.js:0-1009

**Comment by github0013** [comment]
I have another case regarding the same issue.
When you had a machine which would take another copy of the same machine to make parent-child relationship, it would look something like this.

```ts
const create = () => {
  return createMachine({
    initial: "init",
    context: {
      child: null,
    },
    on: {
      SOME_EVENT: {
        actions: ["someAction", sendParent("SOME_EVENT")]
      }
    },
    states: {
      init: {
        entry: assign({
          child: () => spawn(create()),
        }),
      },
    },
  }, {
    actions: {
      someAction: () => {}
    }
  })
}
```

Here I am trying to send `SOME_EVENT` and bubble it up to the top.   It, however, gives an error because it is calling `sendParent("SOME_EVENT")` at the top level.

I am currently doing something like this to avoid calling `sendParent` at the top level.   If `sendParent` just ignores when not possible, I don't have to do the `cond` in the event.

```ts
const create = (isRoot = false) => {
  return createMachine(
    {
      initial: "init",
      context: {
        isRoot,
        child: null,
      },
      on: {
        SOME_EVENT: [
          {
            cond: "isRoot",
            actions: ["someAction"],
          },
          {
            actions: ["someAction", sendParent("SOME_EVENT")],
          },
        ],
      },
      states: {
        init: {
          entry: assign({
            child: () => spawn(create()),
          }),
        },
      },
    },
    {
      actions: {
        someAction: () => {},
      },

      guards: {
        isRoot: (context) => context.isRoot,
      },
    }
  )
}

create(true)
```
